### ISSUE TITLE
feat: run parallel tests with different test contexts

### ISSUE THREAD

**Issue by danielroe** [nuxt-community/module-test-utils#50]
I'd love to be able to initialise `setupTest` several times in the same file or even across files in the test suite.

<details>
<summary>E.g. do something like this</summary>

```js
import { setupTest, createPage, cleanup } from '@nuxtjs/module-test-utils'

describe('module with default options', () => {
  setupTest({
    __dirname,
    browser: true,
    fixture: '../example',
  })

  test('My test', async () => {
    // ...
  })
})

describe('module with other options', () => {
  setupTest({
    __dirname,
    browser: true,
    fixture: '../example',
    config: differentOptions,
  })

  test('My test', async () => {
    // ...
  })
})
```

</details>

If I do this, I get non-determinative behaviour - or errors - even if the test is setup only once in each test spec file. (That is, options *leak* from one test context into another.) Or (if I pass different config objects in different files), I get:
```bash
 TypeError: Cannot read property 'moduleContainer' of undefined

   at node_modules/@nuxtjs/module-test-utils/dist/index.js:286:29
   at fulfilled (node_modules/tslib/tslib.js:112:62)
```

**Comment by pi0** [comment]
According to [jest docs](https://jestjs.io/docs/en/setup-teardown#order-of-execution-of-describe-and-test-blocks) tests in a single file are always executed serially but issue is that inside `setupTest`, we use `beforeAll` which jest executes all of them first (still serially but reordered regardless of describe scope)

Something strange is that I tried changing `beforeAll(() => {})` to `test('build', () => {})` inside `setupTest` but jest still behaves it similar to beforeAll (probably I missed something to try the fix). Only possible way to fix was:

```js
  test('setup', async () => {
    await setupTest({
      __dirname,
      browser: true,
      fixture: '../example',
    })
  }, 60000)
```

**Comment by danielroe** [comment]
> **Once the describe blocks are complete**, by default Jest runs all the tests serially in the order they were encountered in the collection phase, waiting for each to finish and be tidied up before moving on.

That sounds to me like as though where multiple describe blocks in the same file set up different contexts, there will be a conflict. In fact, I've tested further and there is a conflict even when multiple files set up different contexts. The tests at least *run* but one context overwrites the other.

I'm currently wondering if [a custom environment](https://jestjs.io/docs/en/next/configuration#testenvironment-string) might solve the problem and actually add some features/DX.

**Comment by danielroe** [comment]
The simplest solution though is probably to have setupTest return methods already keyed to that context, rather than importing then from the library.

Thoughts?

**Comment by pi0** [comment]
> In fact, I've tested further and there is a conflict even when multiple files set up different contexts. The tests at least run but one context overwrites the other.

There is no shared context between multiple files no matter keeping context in test file or an imported library each file is sandboxed in a worker. The reason multiple setup files are failing is mostly because the setup is also doing a nuxt build which because two runnin in parallel, conflicting on os level... 

```js
// test/shared.js
let state = 0

exports.inc = () => { state ++ }
exports.get = () => state
```

```js
// test/a.test.js
const { get, inc } = require('./shared.js')

describe('A', () => {
  test('inc', () => {
    expect(get()).toBe(0)
    inc()
    expect(get()).toBe(1)
  })
})
```

```js
// test/b.test.js
const { get, inc } = require('./shared.js')

describe('B', () => {
  test('inc', () => {
    expect(get()).toBe(0)
    inc()
    expect(get()).toBe(1)
  })
})
```

**Comment by danielroe** [comment]
Maybe something like #49 could fix then?

**Comment by pi0** [comment]
@danielroe #49 would be perfect if all internals can be instructed to use mock mfs. random build dir is also nice idea for a shorter term. Something else that would be awesome is if we can do snapshot builds similar to new nuxt generate to avoid unnecessary rebuilds when running tests multiple times. We may also introduce a lock-file to fix multi build conflict. 

Also, issue with multiple describes in single file should be fixed by #51 (2.0.0-2) :)

**Comment by pi0** [comment]
Also made #52 :)
